framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,439 @@
+import java.io.Writer;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.element.TypeParameterElement;
+import javax.lang.model.type.TypeMirror;
+import org.checkerframework.javacutil.BugInCF;
+import scenelib.annotations.Annotation;
+import scenelib.annotations.el.AClass;
+import scenelib.annotations.el.AElement;
+import scenelib.annotations.el.AField;
+import scenelib.annotations.el.AMethod;
+import scenelib.annotations.el.AScene;
+import scenelib.annotations.el.ATypeElement;
+import scenelib.annotations.el.AnnotationDef;
+import scenelib.annotations.el.DefCollector;
+import scenelib.annotations.el.DefException;
+import scenelib.annotations.el.InnerTypeLocation;
+import scenelib.annotations.field.AnnotationAFT;
+import scenelib.annotations.field.AnnotationFieldType;
+import scenelib.annotations.field.ArrayAFT;
+import scenelib.annotations.field.BasicAFT;
+import scenelib.annotations.field.ClassTokenAFT;
+import scenelib.annotations.util.Strings;
+
+/**
+ * SceneToStubWriter provides two static methods named {@code write} that write a given {@link
+ * AScene} to a given {@link Writer}, {@link #write(AScene, Map, Map, Set, Writer)}, or filename,
+ * {@link #write(AScene, Map, Map, Set, Writer)}, in stub file format. This class is the equivalent
+ * of {@code IndexFileWriter} from the Annotation File Utilities, but outputs the results in the
+ * stub file format instead of jaif format.
+ *
+ * <p>You can use this writer instead of {@code IndexFileWriter} by passing the {@code
+ * -Ainfer=stubs}.
+ */
+public final class SceneToStubWriter {
+
+    /** How far to indent when writing members of a stub file. */
+    private static final String INDENT = "    ";
+
+    /** The printer where the stub file is written. */
+    private final PrintWriter printWriter;
+
+    /**
+     * A map from the descriptions of ATypeElement_s to the corresponding base Java types, since
+     * AScene_s don't carry that information. See the comment on the {@code basetypes} field of
+     * {@link WholeProgramInferenceScenesHelper} for more information.
+     */
+    private Map<String, TypeMirror> basetypes;
+
+    /**
+     * A map from fully-qualified class names to the TypeElement that represents them. Computed by
+     * {@link WholeProgramInferenceScenes}. Used to output the correct names of generic parameters
+     * to classes, which the stub format requires.
+     */
+    private Map<String, TypeElement> types;
+
+    /**
+     * The fully-qualified name of enums to be written to a stub file.
+     *
+     * <p>The stub parser can't parse an enum that's labeled "class", but {@link AClass} doesn't
+     * specify if a class is an enum. So track which classes are enums.
+     */
+    private Set<String> enumSet;
+
+    /**
+     * Stolen from {@code IndexFileWriter}. Prints a literal; used when printing the arguments of
```

Comment:
Does `stolen` mean that is an exact copy or is it just similar?

Reply:
It is nearly an exact copy. I've made one change, which is to add a case to the end so that long literals are printed with an "L" at the end.

Reply:
I'll update its docs.